AK/BigIntBase.h

```diff
@@ -0,0 +1,595 @@
+// same.
+template<bool is_signed_, size_t bit_size>
+requires(bit_size <= max_big_int_length * word_size) struct StaticStorage {
+    constexpr static size_t static_size = (bit_size + word_size - 1) / word_size;
+    constexpr static bool is_signed = is_signed_;
+
+    // We store integers in little-endian regardless of the host endianness. We use two's complement
+    // representation of negative numbers and do not bother at all if `bit_size % word_size != 0`
+    // (i. e. do not properly handle overflows).
+    NativeWord m_data[static_size];
+
+    constexpr bool is_negative() const
+    {
+        return is_signed_ && m_data[static_size - 1] >> (word_size - 1);
+    }
+
+    constexpr static size_t size()
+    {
+        return static_size;
+    }
+
+    constexpr NativeWord operator[](size_t i) const
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord& operator[](size_t i)
+    {
+        return m_data[i];
+    }
+
+    constexpr NativeWord const* data() const
+    {
+        return m_data;
+    }
+
+    constexpr NativeWord* data()
+    {
+        return m_data;
+    }
+};
+
+// There is no reason to ban u128{0} + 1 (although the second argument type is signed, the value is
+// known at the compile time to be non-negative). In order to do so, we provide overloads in
+// UFixedBigInt which take IntegerWrapper as an argument.
+struct IntegerWrapper {
+    StaticStorage<false, bit_width<int>> m_data;
+
+    consteval IntegerWrapper(int value)
+    {
+        if (value < 0)
+            (new int)&&"Requested implicit conversion of an integer to the unsigned one will underflow.";
+        m_data[0] = static_cast<NativeWord>(value);
+    }
+};
+
+constexpr inline auto get_storage_of(IntegerWrapper value) { return value.m_data; }
+
+template<BuiltInUFixedInt T>
+constexpr StaticStorage<false, bit_width<T>> get_storage_of(T value)
+{
+    if constexpr (sizeof(T) > sizeof(NativeWord)) {
+        static_assert(sizeof(T) == 2 * sizeof(NativeWord));
+        return { static_cast<NativeWord>(value), static_cast<NativeWord>(value >> word_size) };
+    }
+    return { static_cast<NativeWord>(value) };
+}
+
+// ===== Utilities =====
+template<typename T>
+ALWAYS_INLINE constexpr void taint_for_optimizer(T& value)
+{
+    if (!__builtin_is_constant_evaluated()) {
+        asm volatile(""
+                     : "+rm"(value)
+                     :
+                     : "memory");
+    }
+}
+
+ALWAYS_INLINE constexpr NativeWord extend_sign(bool sign)
+{
+    return sign ? max_word : 0;
+}
+
+// FIXME: If available, we might try to use AVX2 and AVX512. We should use something similar to
+//        __builtin_ia32_addcarryx_u64 for GCC on !x86_64 or get PR79713 fixed.
+ALWAYS_INLINE constexpr NativeWord add_words(NativeWord word1, NativeWord word2, bool& carry)
+{
+    if (!__builtin_is_constant_evaluated()) {
+#if ARCH(X86_64)
+        unsigned long long output;
+        carry = __builtin_ia32_addcarryx_u64(carry, word1, word2, &output);
```

Comment:
Regarding performance optimizations: yes, I benchmarked some of these things, and performance difference is noticeable. See #17398, for example.
Regarding `add_words` and `sub_words` specifically, I try to make as clear as I possibly can for the compiler that I want it to use `adc` and `sbb` instead of longer workarounds (and it usually can't do that without a lot of help).